File: ios/Packages/AcquiringKit/Sources/AcquiringCore/CatalogDisplayName.swift
import Foundation

/// Display-only cleanup. Catalog IDs, URLs, and stored names remain unchanged.
public enum CatalogDisplayName {
    public static func title(_ value: String?) -> String {
        format(value) ?? "Unknown Title"
    }

    public static func artist(_ value: String?) -> String {
        if let value {
            let text = clean(value)
            // Names such as blink-182 legitimately contain this hyphen. Without
            // source provenance, preserving it is safer than treating it as a slug.
            if text.range(of: "^[a-z]+-[0-9]+$", options: .regularExpression) != nil {
                return text
            }
        }
        return format(value) ?? "Unknown Artist"
    }

    public static func format(_ value: String?) -> String? {
        guard let value else { return nil }
        let text = clean(value)
        guard !text.isEmpty else { return nil }

        // Separator-only lowercase text is a legacy slug. Do not title-case
        // readable metadata: source styling such as AC/DC and tUnE-yArDs matters.
        guard text.range(of: "^[a-z0-9]+(?:[-_]+[a-z0-9]+)+$", options: .regularExpression) != nil
        else { return text }
        return text.replacingOccurrences(of: "[-_]+", with: " ", options: .regularExpression)
            .capitalized(with: Locale(identifier: "en_US_POSIX"))
    }

    /// Decode display entities and normalize text without changing its spelling or case.
    public static func clean(_ value: String) -> String {
        let decoded = decodeEntities(value).precomposedStringWithCanonicalMapping
        let text = decoded.unicodeScalars.map { scalar -> String in
            if CharacterSet.whitespacesAndNewlines.contains(scalar) { return " " }
            if scalar.value < 0x20 || (0x7f...0x9f).contains(scalar.value) { return "" }
            return String(scalar)
        }.joined()
        return text.components(separatedBy: .whitespacesAndNewlines)
            .filter { !$0.isEmpty }
            .joined(separator: " ")
    }

    private static func decodeEntities(_ value: String) -> String {
        guard let expression = try? NSRegularExpression(pattern: "&(#(?:[xX][0-9a-fA-F]+|[0-9]+)|[a-zA-Z]+);")
        else { return value }
        var text = value
        // A bounded second pass handles doubly escaped source text such as &amp;#39;.
        for _ in 0..<2 {
            let matches = expression.matches(in: text, range: NSRange(text.startIndex..., in: text))
            var changed = false
            for match in matches.reversed() {
                guard let tokenRange = Range(match.range(at: 1), in: text),
                      let entityRange = Range(match.range, in: text)
                else { continue }
                let token = String(text[tokenRange])
                let replacement: String?
                if token.hasPrefix("#") {
                    let isHex = token.lowercased().hasPrefix("#x")
                    let digits = token.dropFirst(isHex ? 2 : 1)
                    replacement = UInt32(digits, radix: isHex ? 16 : 10)
                        .flatMap(Unicode.Scalar.init)
                        .map(String.init)
                } else {
                    replacement = entities[token]
                }
                if let replacement {
                    text.replaceSubrange(entityRange, with: replacement)
                    changed = true
                }
            }
            if !changed { break }
        }
        return text
    }

    private static let entities: [String: String] = [
        "amp": "&", "quot": "\"", "apos": "'", "lt": "<", "gt": ">", "nbsp": " ",
        "lsquo": "‘", "rsquo": "’", "ldquo": "“", "rdquo": "”", "ndash": "–", "mdash": "—",
        "hellip": "…", "eacute": "é", "Eacute": "É", "aacute": "á", "Aacute": "Á",
        "agrave": "à", "Agrave": "À", "egrave": "è", "Egrave": "È", "iacute": "í", "Iacute": "Í",
        "oacute": "ó", "Oacute": "Ó", "uacute": "ú", "Uacute": "Ú", "ntilde": "ñ", "Ntilde": "Ñ",
        "auml": "ä", "Auml": "Ä", "ouml": "ö", "Ouml": "Ö", "uuml": "ü", "Uuml": "Ü",
        "ccedil": "ç", "Ccedil": "Ç", "szlig": "ß", "oslash": "ø", "Oslash": "Ø",
        "aring": "å", "Aring": "Å", "aelig": "æ", "AElig": "Æ"
    ]
}

File: android/app/src/main/java/com/acquiring/android/CatalogDisplayName.kt
package com.acquiring.android

import java.text.Normalizer
import java.util.Locale

/** Display-only cleanup. Catalog IDs, URLs, and stored names remain unchanged. */
object CatalogDisplayName {
    private val blinkStyleArtist = Regex("^[a-z]+-[0-9]+$")
    private val legacySlug = Regex("^[a-z0-9]+(?:[-_]+[a-z0-9]+)+$")
    private val slugSeparators = Regex("[-_]+")
    private val entityPattern = Regex("&(#(?:[xX][0-9a-fA-F]+|[0-9]+)|[a-zA-Z]+);")
    private val posix = Locale.US

    private val entities = mapOf(
        "amp" to "&", "quot" to "\"", "apos" to "'", "lt" to "<", "gt" to ">", "nbsp" to " ",
        "lsquo" to "‘", "rsquo" to "’", "ldquo" to "“", "rdquo" to "”", "ndash" to "–", "mdash" to "—",
        "hellip" to "…", "eacute" to "é", "Eacute" to "É", "aacute" to "á", "Aacute" to "Á",
        "agrave" to "à", "Agrave" to "À", "egrave" to "è", "Egrave" to "È", "iacute" to "í", "Iacute" to "Í",
        "oacute" to "ó", "Oacute" to "Ó", "uacute" to "ú", "Uacute" to "Ú", "ntilde" to "ñ", "Ntilde" to "Ñ",
        "auml" to "ä", "Auml" to "Ä", "ouml" to "ö", "Ouml" to "Ö", "uuml" to "ü", "Uuml" to "Ü",
        "ccedil" to "ç", "Ccedil" to "Ç", "szlig" to "ß", "oslash" to "ø", "Oslash" to "Ø",
        "aring" to "å", "Aring" to "Å", "aelig" to "æ", "AElig" to "Æ"
    )

    fun title(value: String?): String = format(value) ?: "Unknown Title"

    fun artist(value: String?): String {
        if (value != null) {
            val text = clean(value)
            if (blinkStyleArtist.matches(text)) return text
        }
        return format(value) ?: "Unknown Artist"
    }

    fun format(value: String?): String? {
        if (value == null) return null
        val text = clean(value)
        if (text.isEmpty()) return null
        if (!legacySlug.matches(text)) return text
        return slugSeparators.replace(text, " ").split(' ').joinToString(" ") { word ->
            word.replaceFirstChar { ch -> ch.titlecase(posix) }
        }
    }

    fun clean(value: String): String {
        val decoded = Normalizer.normalize(decodeEntities(value), Normalizer.Form.NFC)
        val mapped = buildString {
            decoded.codePoints().forEach { code ->
                val scalar = String(intArrayOf(code), 0, 1)
                when {
                    Character.isWhitespace(code) || Character.isSpaceChar(code) -> append(' ')
                    code < 0x20 || code in 0x7f..0x9f -> Unit
                    else -> append(scalar)
                }
            }
        }
        return mapped.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")
    }

    private fun decodeEntities(value: String): String {
        var text = value
        repeat(2) {
            val matches = entityPattern.findAll(text).toList()
            if (matches.isEmpty()) return text
            var changed = false
            val builder = StringBuilder(text)
            for (match in matches.asReversed()) {
                val token = match.groupValues[1]
                val replacement = if (token.startsWith("#")) {
                    val isHex = token.length >= 2 && token[1].lowercaseChar() == 'x'
                    val digits = token.substring(if (isHex) 2 else 1)
                    digits.toIntOrNull(if (isHex) 16 else 10)?.let { code ->
                        if (code in 0xD800..0xDFFF || !Character.isValidCodePoint(code)) {
                            null
                        } else {
                            String(intArrayOf(code), 0, 1)
                        }
                    }
                } else {
                    entities[token]
                }
                if (replacement != null) {
                    builder.replace(match.range.first, match.range.last + 1, replacement)
                    changed = true
                }
            }
            text = builder.toString()
            if (!changed) return text
        }
        return text
    }
}

val Song.displayTitle: String get() = CatalogDisplayName.title(title)
val Song.displayArtist: String get() = CatalogDisplayName.artist(artist)
val SongBrowseRow.displayTitle: String get() = CatalogDisplayName.title(title)
val SongBrowseRow.displayArtist: String get() = CatalogDisplayName.artist(artist)
